### ISSUE TITLE
Allow caching of password

### ISSUE THREAD

**Issue by edcous** [robinmoisson/staticrypt#116]
I have a Hugo website, using [Netlify](https://www.netlify.com), and would like to know how to not require all pages on the site to ask for the password if it has already been entered on another page. Any help would be greatly appreciated. The following is my Netlify build command
`hugo && npm install && find public -type f -name '*.html' -exec staticrypt {} ****** \\;`
Thanks,
Eli

**Comment by Embraser01** [comment]
Hey, I needed the same thing!

I rewrited a part of the `password_template.html` file to save the pass in the sessionstorage once the user entered the good password.

You can look here: https://github.com/Embraser01/staticrypt/blob/gh-pages/password_template.html

Then for netlify it's really easy!
My `netlify.toml`:
```
[build]
publish = "public"
command = "hugo --gc --minify && ./crypt.sh"

[context.production.environment]
HUGO_VERSION = "0.54.0"
HUGO_ENV = "production"


[context.deploy-preview]
command = "hugo  --gc --minify --buildFuture -b $DEPLOY_PRIME_URL && ./crypt.sh"

[context.deploy-preview.environment]
HUGO_VERSION = "0.54.0"


[context.branch-deploy]
command = "hugo  --gc --minify -b $DEPLOY_PRIME_URL && ./crypt.sh"

[context.branch-deploy.environment]
HUGO_VERSION = "0.54.0"
```

With the following `./crypt.sh`:
```
#!/usr/bin/env bash
set -e

npm install -g https://github.com/Embraser01/staticrypt

find ./public -type f -name "*.html" -exec staticrypt {} $SITE_PASSWORD -o {} \;
```

It will replace all your html files by the encrypted ones. There is still a flash on each pages because I only decrypt after onload but I'll try to improve that quickly.

Also, it's not ideal to save the password as is in the sessionstorage... It would be good to secure it (no idea for now)

**Comment by edcous** [comment]
It didn't work:
Here are my deploy logs:
`9:28:45 AM: Build ready to start
9:28:48 AM: build-image version: 9e0f207a27642d0115b1ca97cd5e8cebbe492f63
9:28:48 AM: build-image tag: v3.3.2
9:28:48 AM: buildbot version: ef8d0929ed0baabafd8bbb7d0b021e1fc24180c0
9:28:48 AM: Fetching cached dependencies
9:28:48 AM: Starting to download cache of 112.0MB
9:28:49 AM: Finished downloading cache in 909.223499ms
9:28:49 AM: Starting to extract cache
9:28:52 AM: Finished extracting cache in 3.364383592s
9:28:52 AM: Finished fetching cache in 4.319441718s
9:28:52 AM: Starting to prepare the repo for build
9:28:53 AM: Preparing Git Reference refs/heads/master
9:28:54 AM: Found netlify.toml. Overriding site configuration
9:28:54 AM: Different build command detected, going to use the one specified in the toml file: 'hugo --gc --minify && ./crypt.sh' versus 'hugo' in the site
9:28:54 AM: Starting build script
9:28:54 AM: Installing dependencies
9:28:54 AM: Started restoring cached node version
9:28:57 AM: Finished restoring cached node version
9:28:57 AM: v10.16.0 is already installed.
9:28:59 AM: Now using node v10.16.0 (npm v6.9.0)
9:28:59 AM: Attempting ruby version 2.6.2, read from environment
9:29:01 AM: Using ruby version 2.6.2
9:29:01 AM: Using PHP version 5.6
9:29:01 AM: Started restoring cached node modules
9:29:01 AM: Finished restoring cached node modules
9:29:01 AM: Started restoring cached go cache
9:29:01 AM: Finished restoring cached go cache
9:29:01 AM: unset GOOS;
9:29:01 AM: unset GOARCH;
9:29:01 AM: export GOROOT='/opt/buildhome/.gimme/versions/go1.12.linux.amd64';
9:29:01 AM: export PATH="/opt/buildhome/.gimme/versions/go1.12.linux.amd64/bin:${PATH}";
9:29:01 AM: go version >&2;
9:29:01 AM: export GIMME_ENV='/opt/buildhome/.gimme/env/go1.12.linux.amd64.env';
9:29:01 AM: go version go1.12 linux/amd64
9:29:01 AM: Installing missing commands
9:29:01 AM: Verify run directory
9:29:01 AM: Executing user command: hugo --gc --minify && ./crypt.sh
9:29:02 AM: Building sites …
9:29:02 AM: Total in 5 ms
9:29:02 AM: Error: Error building site: unmarshal failed: unexpected end of JSON input
9:29:02 AM: Skipping functions preparation step: no functions directory set
9:29:02 AM: Caching artifacts
9:29:02 AM: Started saving node modules
9:29:02 AM: Finished saving node modules
9:29:02 AM: Started saving pip cache
9:29:02 AM: Finished saving pip cache
9:29:02 AM: Started saving emacs cask dependencies
9:29:02 AM: Finished saving emacs cask dependencies
9:29:02 AM: Started saving maven dependencies
9:29:02 AM: Finished saving maven dependencies
9:29:02 AM: Started saving boot dependencies
9:29:02 AM: failed during stage 'building site': Build script returned non-zero exit code: 255
9:29:02 AM: Finished saving boot dependencies
9:29:02 AM: Started saving go dependencies
9:29:02 AM: Finished saving go dependencies
9:29:02 AM: Error running command: Build script returned non-zero exit code: 255
9:29:02 AM: Failing build: Failed to build site
9:29:02 AM: Finished processing build request in 14.251211938s`

**Comment by Embraser01** [comment]
It seems to be an error related to your hugo site (otherwise, it would have printed the results of hugo). Does it compile on your local machine?

**Comment by edcous** [comment]
Re:
> Hey, I needed the same thing!
> 
> I rewrited a part of the `password_template.html` file to save the pass in the sessionstorage once the user entered the good password.
> 
> You can look here: https://github.com/Embraser01/staticrypt/blob/gh-pages/password_template.html
> 
> Then for netlify it's really easy!
> My `netlify.toml`:
> 
> ```
> [build]
> publish = "public"
> command = "hugo --gc --minify && ./crypt.sh"
> 
> [context.production.environment]
> HUGO_VERSION = "0.54.0"
> HUGO_ENV = "production"
> 
> 
> [context.deploy-preview]
> command = "hugo  --gc --minify --buildFuture -b $DEPLOY_PRIME_URL && ./crypt.sh"
> 
> [context.deploy-preview.environment]
> HUGO_VERSION = "0.54.0"
> 
> 
> [context.branch-deploy]
> command = "hugo  --gc --minify -b $DEPLOY_PRIME_URL && ./crypt.sh"
> 
> [context.branch-deploy.environment]
> HUGO_VERSION = "0.54.0"
> ```
> 
> With the following `./crypt.sh`:
> 
> ```
> #!/usr/bin/env bash
> set -e
> 
> npm install -g https://github.com/Embraser01/staticrypt
> 
> find ./public -type f -name "*.html" -exec staticrypt {} $SITE_PASSWORD -o {} \;
> ```
> 
> It will replace all your html files by the encrypted ones. There is still a flash on each pages because I only decrypt after onload but I'll try to improve that quickly.
> 
> Also, it's not ideal to save the password as is in the sessionstorage... It would be good to secure it (no idea for now)

I'm not 100 percent sure on how to set it up, but this is what I did:
added a crypt.sh with the following code:
```
#!/usr/bin/env bash
set -e

npm install -g https://github.com/Embraser01/staticrypt

find ./public -type f -name "*.html" -exec staticrypt {} **** -o {} \\;
```
And, this is my netlify.toml:
```
[build]
  publish = "public"
  command = "hugo && npm install && ./crypt.sh"
[build.environment]
  HUGO_VERSION = "0.55.6"
  HUGO_THEME = "hugo-hero-theme"
  HUGO_BASEURL = "/"
```

**Comment by kevincolten** [comment]
I got it working using this command
` set -e && find ./_site -type f -name '*.html' -exec staticrypt -f _javascripts/password_template.html {} $PASSWORD -o {} \\;
`

and by tweaking the password template script, replacing
```js
document.getElementById('staticrypt-form').addEventListener('submit', function(e) {
    e.preventDefault();
    var passphrase = document.getElementById('staticrypt-password').value,
        encryptedMsg = '{encrypted}',
        encryptedHMAC = encryptedMsg.substring(0, 64),
        encryptedHTML = encryptedMsg.substring(64),
        decryptedHMAC = CryptoJS.HmacSHA256(encryptedHTML, CryptoJS.SHA256(passphrase).toString()).toString();
    if (decryptedHMAC !== encryptedHMAC) {
        alert('Bad passphrase!');
        return;
    }
    var plainHTML = decrypt(encryptedHTML, passphrase);
    document.write(plainHTML);
    document.close();
});
```
with

```js
function insertPlainHTML(password) {
    encryptedMsg = '{encrypted}',
    encryptedHMAC = encryptedMsg.substring(0, 64),
    encryptedHTML = encryptedMsg.substring(64),
    decryptedHMAC = CryptoJS.HmacSHA256(encryptedHTML, CryptoJS.SHA256(password).toString()).toString();

    if (decryptedHMAC !== encryptedHMAC) {
        // alert('Bad password!');
        localStorage.removeItem('password');
        return;
    }
    localStorage.setItem('password', btoa(password));
    document.write(decrypt(encryptedHTML, password));
    document.close();
}

window.addEventListener('DOMContentLoaded', function() {
    if (localStorage.getItem('password')) {
        insertPlainHTML(atob(localStorage.getItem('password')))
    } else {
        document.getElementById('staticrypt-form').addEventListener('submit', function(e) {
            e.preventDefault();
            insertPlainHTML(document.getElementById('staticrypt-password').value);
        });
    }
});
```

**Comment by bsde1234** [comment]
Stupid question do I need to run `./crypt.sh` locally or somewahere in netlify or in hugo template?

**Comment by imshuffling** [comment]
Any examples of this working for gatsby?

Asking for me a friend :)

**Comment by imshuffling** [comment]
^ Got it working

```
find ./public -type f -name \"*.html\" -exec npx staticrypt -f \"./src/components/password_template.html\" {} **** -o {} \\;
```

**Comment by edcous** [comment]
Anyone know how to do this for Jekyll?

**Comment by edcous** [comment]
> Stupid question do I need to run `./crypt.sh` locally or somewahere in netlify or in hugo template?

In netlify, in your build command

**Comment by staticman-cosmos** [comment]
This is very awesome @Embraser01  -- has anybody made any progress in removing the flash of the password box before decryption? I've just start looking at the code a bit but hoping to find some way to replace that with a loading spinner or similar.

**Comment by 888ba8** [comment]
Hi,

I am using the following Gatsby template https://jamtemplates.com/templates/gatsby-yellowcake and Deployed it to netlify via a private Gitlab repository.

I am using the following tutorial https://medium.com/@j0hnm4r5/password-protecting-a-website-with-the-free-tier-of-netlify-and-publishing-with-a-single-npm-e509af9f108f to password protect a single file public/index.html

and it works great.

What I did was I replace whatever was in my **package.json** `"scripts": {` by what the tutorial mentions.

Then I changed my **netlify.toml** command to `command = "npm run publish"`

and under **https://app.netlify.com/sites/elements-development/settings/deploys** I also changed the build command to `npm publish`

However, how can I password protect each individual page / the whole website please?

Kind regards,

888

**Comment by 888ba8** [comment]
Can you help me with the following question:

https://stackoverflow.com/questions/65129417/gatsby-browser-remembers-password-jekyll-browser-does-not-remember-password

**Comment by robinmoisson** [comment]
Hey all, thanks for the issue and discussion. This is closed in #124 , which adds a "Remember me" checkbox. The passphrase is stored hashed + salted in localStorage, meaning it's a slightly bigger attack surface for decryting the encrypted page, but should mitigate password reuse leakage (see discussion in #120 ).

@staticman-cosmos regarding the flash of the password box - you could modify the `password_template` file to show a spinner by default and hide the whole password box with a `hidden` class, and in the `onload` function you check:
- if there is a stored passphrase, decrypt and show that
- if there isn't (or it's wrong), show the password box instead

This means that you might have very quick flash of spinner before the password box shows for a new user.
